### ISSUE TITLE
libai 开发 check list （持续更新）

### ISSUE THREAD

**Issue by dangkai4u** [Oneflow-Inc/libai#38]
这里汇总libai项目需要支持的所有功能，以及它们是否完成，方便大家了解当前的开发进度，知道框架欠缺什么，可以主动认领任务或补齐特性。同时，汇总出来已实现的功能，之后可以根据此编写使用文档：

### 模块抽象
- [x] linear
- [x] multihead attention
- [x] layer norm
- [x] mlp
- [x] lm_logits
- [x] vocab_embedding，embedding，position embedding
- [x] transformer layer
- [x] activation
- [ ] cross entropy：由于框架内的损失函数都是常见的，且它仅实现损失计算，不涉及logging记录的功能，我（党凯）认为，这里用oneflow内置的损失函数即可，ignore_index和label smoothing这两个常用特性，如果内置损失函数不支持，应该在那里修改，不要交给用户。
- [ ] attention with relational position embedding

### 模型
- [ ] BERT：doing
- [ ] GPT：doing
- [ ] T5：doing
- [ ] VIT：doing
- [ ] realm：结合检索的语言模型

### 配置系统功能
- [x] yacs-based + 命令行参数
- [x] LazyCall
- [x] 注册机制

### trainer
- [ ] 构建graph
- [ ] 构建optimizer
- [ ] 构建lr scheduler
- [ ] 训练、验证模型
- [x] 日志记录
- [x] 打印训练信息metric
- [ ] 保存模型
- [ ] 加载模型

### 数据加载
- [ ] bert dataset
- [ ] gpt dataset
- [ ] t5 dataset
- [ ] vit dataset
- [ ] realm dataset
- [ ] 数据集切分
- [ ] 数据集合并

### 优化器、调度器
- [ ] adam（fused）
- [ ] adamw（fused）
- [ ] sgd（fused）
- [ ] LAMB（fused）
- [ ] LARS
- [ ] poly decay
- [ ] cosine decay
- [ ] linear decay
- [ ] inverse root sqrt decay
- [ ] multistep decay

### 通用特性
- [ ] 模型并行下的2D、序列并行等尝试
- [ ] generator构造
- [ ] 下游任务结果复现

### eager下的特性
- [ ] amp训练
- [ ] zero优化
- [ ] moe机制
- [ ] checkpointing

### graph下的特性
- [x] amp训练
- [ ] zero优化
- [ ] moe机制
- [x] checkpointing

## 第三阶段：部署、量化、蒸馏、实现API接口
- [ ] 抽象出知识蒸馏框架
- [ ] 实现模型量化
- [ ] 允许将训练出的模型对接huggingface、或者转ONNX用于部署
- [ ] 提供API接口，用户通过API调用模型，做zero shot learning或few shot learning
